library/common/main_interface.h

```diff
@@ -1,12 +1,141 @@
 #pragma once
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
 
 // NOLINT(namespace-envoy)
 
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef int envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP request. Valid only for the duration of the request and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_request_t;
```

Comment:
Assuming the intent for this is to have the buffer/lifecycle managed by the shim (or platform) library, and pass an `envoy_request_t` as a pointer to the underlying C library?

Reply:
There are several ways to do this. The way @goaway and I were discussion with this "handle" was to actually let the c/c++ implementation keep track of all the request (e.g on a static table) via these integer handles as IDs.

Reply:
Do we need to expose the non-streaming API from core? I know we have discussed this a bit in the past, but my general feeling is that we should propagate streaming as far up the stack as possible as it will avoid issues and assumptions later.

Reply:
> The way @goaway and I were discussion with this "handle" was to actually let the c/c++ implementation keep track of all the request (e.g on a static table) via these integer handles as IDs.

Got it. That makes sense, though I think it's worth documenting here since it's not immediately obvious by the type as to what this represents.

> Do we need to expose the non-streaming API from core? I know we have discussed this a bit in the past, but my general feeling is that we should propagate streaming as far up the stack as possible as it will avoid issues and assumptions later.

I don't think we _need_ to expose the non-streaming API. However, this does mean that we'll be exposing unary calls from the Swift/Kotlin libraries, and writing (minimal?) bridging logic back to the C streaming APIs in both places. Is that the intent?

Reply:
yep, that is the intent. Will respond on main thread.